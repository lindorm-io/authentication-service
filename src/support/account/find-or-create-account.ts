import { Account } from "../../entity";
import { IKoaAuthContext } from "../../typing";
import { InvalidPermissionError } from "../../error";
import { requestEnsureIdentity } from "../../axios";
import { isLocked } from "@lindorm-io/jwt";

export const findOrCreateAccount = (ctx: IKoaAuthContext) => async (email: string): Promise<Account> => {
  const { repository } = ctx;

  const account = await repository.account.findOrCreate({ email });

  if (isLocked(account.permission)) {
    throw new InvalidPermissionError();
  }

  if (!account.identityLinked) {
    await requestEnsureIdentity(ctx)(account);

    account.identityLinked = true;

    await repository.account.update(account);
  }

  return account;
};
