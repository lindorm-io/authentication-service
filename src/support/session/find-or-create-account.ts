import { Account } from "../../entity";
import { IAuthContext } from "../../typing";
import { RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { Permission } from "../../enum";
import { InvalidPermissionError } from "../../error";

export const findOrCreateAccount = (ctx: IAuthContext) => async (email: string): Promise<Account> => {
  const { repository } = ctx;

  try {
    const account = await repository.account.find({ email });

    if (account.permission === Permission.LOCKED) {
      throw new InvalidPermissionError();
    }

    return account;
  } catch (err) {
    if (err instanceof RepositoryEntityNotFoundError) {
      const account = new Account({ email });
      account.create();

      return repository.account.create(account);
    }

    throw err;
  }
};
