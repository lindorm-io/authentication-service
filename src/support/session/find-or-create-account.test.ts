import MockDate from "mockdate";
import { RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { findOrCreateAccount } from "./find-or-create-account";
import { getMockRepository } from "../../test/mocks";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("findOrCreateAccount", () => {
  let getMockContext: any;

  beforeEach(() => {
    getMockContext = () => ({
      repository: getMockRepository(),
    });
  });

  test("should return an existing account", async () => {
    await expect(findOrCreateAccount(getMockContext())("existing")).resolves.toMatchSnapshot();
  });

  test("should create a new account", async () => {
    const context = getMockContext();
    const ctx = {
      ...context,
      repository: {
        ...context.repository,
        account: {
          ...context.repository.account,
          find: jest.fn(() => {
            throw new RepositoryEntityNotFoundError({}, {});
          }),
        },
      },
    };

    await expect(findOrCreateAccount(ctx)("created")).resolves.toMatchSnapshot();
  });

  test("should reject with error", async () => {
    const context = getMockContext();
    const ctx = {
      ...context,
      repository: {
        ...context.repository,
        account: {
          ...context.repository.account,
          find: jest.fn(() => {
            throw new Error("error");
          }),
        },
      },
    };

    await expect(findOrCreateAccount(ctx)("reject")).rejects.toStrictEqual(expect.any(Error));
  });
});
