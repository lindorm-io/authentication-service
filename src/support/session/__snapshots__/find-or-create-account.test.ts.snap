// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findOrCreateAccount should create a new account 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "created",
  "_events": Array [
    Object {
      "created": 2020-01-01T07:00:00.000Z,
      "name": "account_created",
      "payload": Object {
        "created": 2020-01-01T07:00:00.000Z,
        "email": "created",
        "otp": Object {
          "signature": null,
          "uri": null,
        },
        "password": Object {
          "signature": null,
          "updated": null,
        },
        "permission": "user",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_otp": Object {
    "signature": null,
    "uri": null,
  },
  "_password": Object {
    "signature": null,
    "updated": null,
  },
  "_permission": "user",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`findOrCreateAccount should return an existing account 1`] = `
Account {
  "_created": 2020-01-01T07:00:00.000Z,
  "_email": "existing",
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_otp": Object {
    "signature": null,
    "uri": null,
  },
  "_password": Object {
    "signature": null,
    "updated": null,
  },
  "_permission": "user",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
